design/open/template-server/template-server.md

```diff
@@ -0,0 +1,77 @@
+# Template Server
+
+## About
+
+The template server adds a simple generic templating capability to the
+lakeFS web server.  It adds server-side support for administrators and
+developers to serve templated web objects securely.
+
+Among the planned applications:
+
+* Serve capabilities to clients.  Serving capabilities from a template
+  lets developers add them easily but allows administrators to control
+  what is actually served.
+* Serve prepopulated config files for popular setups.  For instance we
+  can serve `.lakectl` and Hadoop/Spark config files.  These sometimes
+  need to be modified, for instance when lakeFS does not know the name
+  of its endpoint.
+
+  This can be an important part of for time-to-value.  Using templates
+  we offer admins some flexibility in configuration: for instance they
+  might define templates with different values, and clients can select
+  which to receive by path.
+  
+## Template format
+
+We use Golang [text/template][text/template] to expand most templates.
+For browsers (based on filename, e.g. `*.html.tt`) we instead must use
+[html/template][html/template] for safety.
+
+The template is read by lakeFS from a specific (configured) prefix, by
+default `lakefs://lakefs-templates/main`.  IAM authorizes users to see
+this template, at which point expansion can start.  We will supply the
+following replacements (at least):
+
+| template function | value source              | IAM permissions                              |
+|:------------------|:--------------------------|:---------------------------------------------|
+| config            | lakeFS configuration[^1]  | `arn:lakefs:template:::config/path/to/field` |
+| object            | (small) lakeFS object     | IAM read permission for that object          |
+| querystring       | URL query string[^2]      | (none)                                       |
+| contenttype       | none; set `Content-Type:` | (none)                                       |
+
+Adding a freeform dictionary to our config will allow admins to set up
+any needed configuration.
+
+We shall also supply the user object in `.user`.  We will probably add
+new functions or objects.
+
+## Examples
+
+(All examples subject to change as we define object format, and decide
+what functions are allowed.)
+
+Set a template that looks like this:
+
+```conf
+spark.hadoop.fs.lakefs.impl=io.lakefs.LakeFSFileSystem
+spark.hadoop.fs.lakefs.access_key={{.user.credentials.access_key.id}}
+spark.hadoop.fs.lakefs.secret_key={{.user.credentials.access_key.secret}}
```

Comment:
Not sure it's a good idea to expose the secrets through templating

Reply:
Thanks for clarifying about these being taken from the request.

So yes, let's discuss returning the underlying storage credentials from the configuration: I'm not referring specifically to templating - I think in general we shouldn't return them from any API endpoint.